pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,255 @@
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A time tracker class to measure time spent on different sections of PMD analysis.
+ * The class is thread-aware, allowing to differentiate CPU and wall clock time.
+ * 
+ * @author Juan Martín Sotuyo Dodero
+ */
+public final class TimeTracker {
+
+    private static boolean trackTime = false;
+    private static long wallClockStart = -1;
+    private static final ThreadLocal<Queue<TimerEntry>> TIMER_ENTRIES;
+    private static final ConcurrentMap<TimedOperation, TimedResult> ACCUMULATED_RESULTS = new ConcurrentHashMap<>();
+    
+    static {
+        TIMER_ENTRIES = new ThreadLocal<Queue<TimerEntry>>() {
+            @Override
+            protected Queue<TimerEntry> initialValue() {
+                final Queue<TimerEntry> queue = Collections.asLifoQueue(new LinkedList<TimerEntry>());
+                queue.add(new TimerEntry(TimedOperationCategory.UNACCOUNTED, null));
+                return queue;
+            }
+        };
+    }
+    
+    private TimeTracker() {
+        throw new AssertionError("Can't instantiate utility class");
+    }
+    
+    /**
+     * Starts global tracking. Allows tracking operations to take place and starts the wall clock.
+     * Must be called once PMD starts if tracking is desired, no tracking will be performed otherwise.
+     */
+    public static void startGlobalTracking() {
+        wallClockStart = System.currentTimeMillis();
+        trackTime = true;
+        initThread(); // init main thread
+    }
+    
+    /**
+     * Stops global tracking. Stops the wall clock. All further operations will be treated as NOOP.
+     * @return The timed data obtained through the run.
+     */
+    public static TimingReport stopGlobalTracking() {
+        if (!trackTime) {
+            return null;
+        }
+        
+        finishThread(); // finish the main thread
+        trackTime = false;
+        
+        // Fix UNACCOUNTED metric (total time is meaningless)
+        final TimedResult unaccountedResult = ACCUMULATED_RESULTS.get(
+                new TimedOperation(TimedOperationCategory.UNACCOUNTED, null));
+        unaccountedResult.totalTime.set(unaccountedResult.selfTime.get());
+        
+        return new TimingReport(System.currentTimeMillis() - wallClockStart, ACCUMULATED_RESULTS);
+    }
+    
+    /**
+     * Initialize a thread, starting to track it's own time.
+     */
+    public static void initThread() {
+        if (!trackTime) {
+            return;
+        }
+        
+        TIMER_ENTRIES.get(); // Just make sure it's initialized
+    }
+    
+    /**
+     * Finishes tracking a thread.
+     */
+    public static void finishThread() {
+        if (!trackTime) {
+            return;
+        }
+        
+        // if using a mono-thread, we may not be empty...
```

Comment:
Why is this? Seems like some kind of off-by-one bug.  Shouldn't the starter always call finish?

What if it were 7 and not 1?  Anything but 0 seems like a potential API misuse of some sort...

Reply:
When using mono-thread, the main thread calls the `PmdRunnable`, so the thread local contains the same instance. In that scenario, `startThread` retrieved a non-empty local, and added nothing to it, so we shouldn't remove anything from it either.

I originally had the init push an unaccounted state manually and pop it back on finish, but I had another problem with that I can't recall now... for sure it would have a call count for unaccounted higher than expected

Reply:
I'm assuming `startThread` the same as `initThread`.

Is this related to the nesting rollup somehow?  I'm still not how the bookkeeping gets wonky in the monothread case, or why it causes a problem.

If you think this is right that's ok with me.  I don't have the cycles to fire up the code to better understand right now.